### DISCUSSION TITLE
https CURL_MAX_WRITE_SIZE

### DISCUSSION THREAD

**Discussion by ddxxsx** [curl/curl#12725] [Q&A] [answered]
I'm using libcurl to download files using https protocol. Now I'm certain I have set the CURL_MAX_WRITE_SIZE to larger than 4096, however, function"write_data()" which I set in CURLOPT_WRITEFUNCTION seems always return a buffer_length close to 4096.
I suppose it is because of the Openssl buffer_size limit. And I have two questions.
1. Is there a way to increase the buffer_length in "write_data()"
2. Is the maximum value for buffer size 16k due to TSL protocol?

**Comment by bagder** [comment, marked as answer]
> Is there a way to increase the buffer_length in "write_data()"

There is no such 4K limit anywhere in curl. If you get callbacks called with 4K data, it is because libcurl received that amount from the stack and wants to dispose of it as fast as possible to the application.

> Is the maximum value for buffer size 16k due to TSL protocol?

You mean the selected value for `CURL_MAX_WRITE_SIZE` ? No, it was arbitrarily picked in the early 2000s after some basic experiments. It has nothing to do with TLS.

**Comment by ddxxsx** [comment]
I changed the selected value for CURL_MAX_WRITE_SIZE  and set it to be 128K, and download files twice, the first time I use http protocol and the callback has data close to 128K. The second time I use https protocol and the callback has data limited to 4K.
Later I discover the problem only exists on version before curl-7_87_0(Mine was 7.35) due to this commit  https://github.com/curl/curl/commit/dafdb20a26d0c890e83dea61a104b75408481ebd#diff-76e92705200fa8317ac3f635fd17d0ffe976a42e304edfd6c5ce024857a14ca0
which changes the Implement of function "data_pending()" in file "transfer.c". After which  data_pending() not only check if ssl has data_pending, but also the whole connection.
I believe this to be the reason for my question, thank you for your assistance

**Comment by icing** [comment]
With a curl v8.3.0 or newer, you can easily trace what is happending at the SSL layer. For example:

```
> curl  -v --http1.1 --trace-config ids,time,ssl https://curl.se/download/curl-8.5.0.tar.gz -o /dev/null
```

will show you the amount of data returned by the SSL layer or receive and the calls it does against the socket and what that returned. In general, the pattern is:

```
08:42:05.651684 [0-0] * [SSL] ossl_bio_cf_in_read(len=16401) -> 2687, err=0
08:42:05.651710 [0-0] * [SSL] ossl_bio_cf_in_read(len=13714) -> -1, err=81
08:42:05.651733 [0-0] * [SSL] cf_recv(len=17907) -> -1, 81
08:42:05.651831 [0-0] * [SSL] ossl_bio_cf_in_read(len=13714) -> 1420, err=0
08:42:05.651858 [0-0] * [SSL] ossl_bio_cf_in_read(len=12294) -> -1, err=81
08:42:05.651881 [0-0] * [SSL] cf_recv(len=17907) -> -1, 81
08:42:05.651948 [0-0] * [SSL] ossl_bio_cf_in_read(len=12294) -> 2840, err=0
08:42:05.651975 [0-0] * [SSL] ossl_bio_cf_in_read(len=9454) -> -1, err=81
08:42:05.651998 [0-0] * [SSL] cf_recv(len=17907) -> -1, 81
08:42:05.652399 [0-0] * [SSL] ossl_bio_cf_in_read(len=9454) -> 1420, err=0
08:42:05.652425 [0-0] * [SSL] ossl_bio_cf_in_read(len=8034) -> -1, err=81
08:42:05.652448 [0-0] * [SSL] cf_recv(len=17907) -> -1, 81
08:42:05.652621 [0-0] * [SSL] ossl_bio_cf_in_read(len=8034) -> 1420, err=0
08:42:05.652647 [0-0] * [SSL] ossl_bio_cf_in_read(len=6614) -> -1, err=81
08:42:05.652670 [0-0] * [SSL] cf_recv(len=17907) -> -1, 81
08:42:05.652973 [0-0] * [SSL] ossl_bio_cf_in_read(len=6614) -> 2840, err=0
08:42:05.653001 [0-0] * [SSL] ossl_bio_cf_in_read(len=3774) -> -1, err=81
08:42:05.653025 [0-0] * [SSL] cf_recv(len=17907) -> -1, 81
08:42:05.653085 [0-0] * [SSL] ossl_bio_cf_in_read(len=3774) -> 1420, err=0
08:42:05.653112 [0-0] * [SSL] ossl_bio_cf_in_read(len=2354) -> -1, err=81
08:42:05.653135 [0-0] * [SSL] cf_recv(len=17907) -> -1, 81
08:42:05.653545 [0-0] * [SSL] ossl_bio_cf_in_read(len=2354) -> 2354, err=0
08:42:05.653576 [0-0] { [1 bytes data]
08:42:05.653612 [0-0] * [SSL] cf_recv(len=17907) -> 16384, 0
```
Here you see a TLS record being read in several chunks. The network `ossl_bio_cf_in_read()` is called by OpeSSL again and again, until the data for the current TLS record is complete (err=81 means EAGAIN), then it returns it as mentioned in `[SSL] cf_recv(len=17907) -> 16384, 0`.

This means that `curl` gets the DATA in the chunk sizes that the *server* chose for its TLS packets. Currently, we do not buffer those extra and directly forward them to the application. If your `https:` server sends in 4k TLS records, you will see those.

In plain `http:` we receive at max buffer size from the socket, and if the serrver is fast enough, you will get the bufffer filled.

Now: the question is: would it be beneficial to try to read more from OpenSSL until the read buffer is full or an EAGAIN happened? CC: @bagder. A question of preference "immediate" delivery vs. "as large as possible".

**Reply by bagder** [reply]
> A question of preference "immediate" delivery vs. "as large as possible".

I think that if we can prove significant enough performance benefits we should consider changing this. Otherwise not.

**Reply by icing** [reply]
Scorecard differences between master and #12801. Notice that scorecard uses `-o /dev/null`, so any overhead encountered in client writes is neglible. 

Single transfer see no difference. They are called often enough without interference. Parallel transfers are better with less invocations. I would assume the effect on an app actually processing the received data being even greater.

```
> python3 tests/http/scorecard.py -d h1 --httpd

*master*
HTTP/1.1 in curl 8.6.0-DEV (x86_64-apple-darwin23.2.0) libcurl/8.6.0-DEV quictls/3.1.4 zlib/1.2.12 brotli/1.1.0 zstd/1.5.5 c-ares/1.21.0 libidn2/2.3.4 libpsl/0.21.2 libssh2/1.11.0 nghttp2/1.59.0-DEV ngtcp2/1.1.0 nghttp3/1.1.0 librtmp/2.3
Downloads
  Server       Size   single(1x) [cpu/rss]          serial(50x) [cpu/rss]         parallel(50x) [cpu/rss]                Errors
  httpd         1MB      22 MB/s [61.4%/2MB]           400 MB/s [82.4%/7MB]           340 MB/s [95.1%/13MB]               -
  httpd        10MB     183 MB/s [69.7%/4MB]           872 MB/s [95.7%/8MB]          1042 MB/s [99.6%/16MB]               -
  httpd       100MB     676 MB/s [94.2%/7MB]           948 MB/s [97.3%/8MB]          1278 MB/s [100.0%/16MB]              -

*#12801*
HTTP/1.1 in curl 8.6.0-DEV (x86_64-apple-darwin23.2.0) libcurl/8.6.0-DEV quictls/3.1.4 zlib/1.2.12 brotli/1.1.0 zstd/1.5.5 c-ares/1.21.0 libidn2/2.3.4 libpsl/0.21.2 libssh2/1.11.0 nghttp2/1.59.0-DEV ngtcp2/1.1.0 nghttp3/1.1.0 librtmp/2.3
Downloads
  Server       Size   single(1x) [cpu/rss]          serial(50x) [cpu/rss]         parallel(50x) [cpu/rss]                Errors
  httpd         1MB      22 MB/s [61.1%/2MB]           397 MB/s [81.9%/8MB]           341 MB/s [94.4%/13MB]               -
  httpd        10MB     180 MB/s [70.0%/4MB]           871 MB/s [95.8%/9MB]          1138 MB/s [99.4%/20MB]               -
  httpd       100MB     674 MB/s [94.4%/7MB]           948 MB/s [97.3%/9MB]          1475 MB/s [99.9%/20MB]               -
``
